// @mikro-orm/mongo-highlighter/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOKEN_TYPE_TO_HIGHLIGHT = exports.HighlightSubject = exports.TokenType = void 0;
var TokenType;
(function (TokenType) {
    TokenType[TokenType["WHITESPACE"] = 0] = "WHITESPACE";
    TokenType[TokenType["WORD"] = 1] = "WORD";
    TokenType[TokenType["QUOTE"] = 2] = "QUOTE";
    TokenType[TokenType["BACKTICK_QUOTE"] = 3] = "BACKTICK_QUOTE";
    TokenType[TokenType["BOUNDARY"] = 4] = "BOUNDARY";
    TokenType[TokenType["NUMBER"] = 5] = "NUMBER";
    TokenType[TokenType["ERROR"] = 6] = "ERROR";
    TokenType[TokenType["VARIABLE"] = 7] = "VARIABLE";
    TokenType[TokenType["LITERAL"] = 8] = "LITERAL";
})(TokenType = exports.TokenType || (exports.TokenType = {}));
var HighlightSubject;
(function (HighlightSubject) {
    HighlightSubject["BOUNDARY"] = "boundary";
    HighlightSubject["WORD"] = "word";
    HighlightSubject["BACKTICK_QUOTE"] = "backtickQuote";
    HighlightSubject["QUOTE"] = "quote";
    HighlightSubject["NUMBER"] = "number";
    HighlightSubject["VARIABLE"] = "variable";
    HighlightSubject["LITERAL"] = "literal";
})(HighlightSubject = exports.HighlightSubject || (exports.HighlightSubject = {}));
exports.TOKEN_TYPE_TO_HIGHLIGHT = {
    [TokenType.BOUNDARY]: HighlightSubject.BOUNDARY,
    [TokenType.WORD]: HighlightSubject.WORD,
    [TokenType.BACKTICK_QUOTE]: HighlightSubject.BACKTICK_QUOTE,
    [TokenType.QUOTE]: HighlightSubject.QUOTE,
    [TokenType.NUMBER]: HighlightSubject.NUMBER,
    [TokenType.VARIABLE]: HighlightSubject.VARIABLE,
    [TokenType.LITERAL]: HighlightSubject.LITERAL,
};
